### ISSUE TITLE
absolutely unclear documentation

### ISSUE THREAD

**Issue by MODE83-PROJET-DEV** [flutter/website#11127]
### Page URL

https://docs.flutter.dev/platform-integration/web/initialization/

### Page source

https://github.com/flutter/website/tree/main/src/content/platform-integration/web/initialization.md

### Describe the problem

In the doc, I would like to know more about how to manage the worker service and in particular forcing the update of the app via the worker.

1: the code examples are not precise enough and lack a lot of information.

2: Where is how I should write my JS code on the project's web folder

3: Why don't I have clear information to manage the worker service?

Thank you for your attention.
Sincerely

### I would like to fix this problem.

- [ ] I will try and fix this problem on docs.flutter.dev.

**Comment by sfshaza2** [comment]
@johnpryan, do you have any sage advice?

**Comment by MODE83-PROJET-DEV** [comment]
@sfshaza2 Yes, how to reload the worker service, via your methods? and then how can I retrieve the version code of my application contained in pubspec.yaml and compare it with the version code loaded on the browser.Yes, how to reload the worker service, via your methods? and then how can I retrieve the version code of my application contained in pubspec.yaml and compare it with the version code loaded on the browser.

Thanks

**Comment by MODE83-PROJET-DEV** [comment]
Searching deeper, I came across this:

https://github.com/flutter/flutter/issues/149031

It would be good if the problem were addressed as a priority, because it seriously harms the user experience.

Sincerely

**Comment by sfshaza2** [comment]
@kevmoo, can you weigh in?

**Comment by sfshaza2** [comment]
@MODE83-PROJET-DEV, access to which source code is restricted? Do you mean access to Chrome source? Because, yes, indeed.

**Comment by kevmoo** [comment]
@sigmundch is a better expert here.

It's tricky to say the least...

**Comment by sigmundch** [comment]
Sorry I don't believe I have context here.

Sounds like the question here is about customizing flutter web's initialization worker, specifically the logic that involves web workers? Maybe @ditman or @yjbanov have more context here?

**Comment by xvrh** [comment]
The ServiceWorker "Offline first" offered by default by Flutter Web is very confusing.

When we push a new version of our app, in Safari, no matter how many times you press the refresh button in your browser, you won't get the new version. You'll be stuck on the previous version for a day 😳.

The documentation should clearly explains this behaviour and provide some alternative configurations of the ServiceWorker to unblock situations like that. And show an example of a "Reload" button to force the refresh when a new version is detected.

**Comment by MODE83-PROJET-DEV** [comment]
@xvrh I agree with you, the documentation needs to be improved. 
And also the Flutter Worker service which must also compare the version codes and force the application to be updated each time the version changes. And for most of my users, this bothers them a lot, because they don't have the reflex to clear their browser caches.

@sfshaza2 Yes indeed, I don't have access to anything to do an investigation. thank you for your response. However, the problem is much more complex than I thought. And there are already similar discussions about this here: 

https://github.com/flutter/flutter/issues/149031

Expertise is needed on this issue, I am far from understanding the situation.
In any case, thank you very much for your attention, it's cool.
Sincerely

**Comment by MODE83-PROJET-DEV** [comment]
Good morning, 

I have a local test in a separate environment with only the bare essentials apart from the test server under Linux and the production which also runs under Linux. 

Ultimately this comes from a caching problem on the server side, because locally in my environment, there is indeed a change in version of the application.

Tested on Chrome, Firefox and Edge.

On my production server, I had to add this to the htaccess:

```
<IfModule mod_headers.c>
    Header set Access-Control-Allow-Origin "*"
    Header set Cache-Control "no-cache, no-store, must-revalidate"
    Header set Pragma "no-cache"
    Header set Expires "0"
</IfModule>

<IfModule mod_expires.c>
  ExpiresActive On
  ExpiresByType text/html "access plus 1 days"
  ExpiresByType image/gif "access plus 1 days"
  ExpiresByType image/jpeg "access plus 1 days"
  ExpiresByType image/png "access plus 1 days"
  ExpiresByType text/css "access plus 1 days"
  ExpiresByType text/javascript "access plus 1 days"
  ExpiresByType application/javascript "access plus 1 days"
  ExpiresByType application/x-javascript "access plus 1 days"
  ExpiresByType application/pdf "access plus 1 days"
  ExpiresByType application/x-shockwave-flash "access plus 1 days"
  ExpiresByType image/x-icon "access plus 1 days"
  ExpiresDefault "access plus 1 days"
</IfModule>
```

Now, is this a solution...? In any case now when I make a modification it is there, as soon as we reload the page.
For now I'm going to leave it like that and see later if there are more attractive solutions.

**Comment by ditman** [comment]
There have been discussions on the team to replace the default service worker that `flutter build` generates by some templating so people can create their own service workers using Workbox (a framework for service workers), which is (very likely) much better than anything we do now!

See: https://developer.chrome.com/docs/workbox

**Comment by ditman** [comment]
@MODE83-PROJET-DEV If your users are suffering with the default flutter service worker, you could potentially create your own that completely disables any service worker on your web, so your users only have to refresh one final time.

Then the standard browser behavior would control caching normally (respecting your server-side configuration).

**Comment by sfshaza2** [comment]
Yowza. This is definitely beyond the sort of thing that we would typically document on docs.flutter.dev.

So, I'm wondering, @ditman, @MODE83-PROJET-DEV... Is this issue "enough"? In other words, if folks are searching for help on this, they will find the issue, learn about the complexities, *and* the current workaround, which is to destroy Flutter's default behavior for service worker initialization?

It sounds like this issue will eventually be solved...

It would be nice if we had a blog post about this.  :)

**Comment by MODE83-PROJET-DEV** [comment]
Thank you all for your help, for me strangely the modification of the .htaccess parameters forces the worker service to update itself... But for all that I would use your solutions to apply the destruction of the worker service on each loading.

Thank you for all your information.

@sfshaza2 It is true that with the experience gathered here, there is enough to document this problem or open a blog.
However, I don't see where to vote? and why?

**Comment by ditman** [comment]
@sfshaza2 @MODE83-PROJET-DEV I think this needs to be taken to the flutter/flutter repo so we can decide what to do with the current version of the Service Worker!

As I mentioned on the issue above, we might want to remove the default service worker altogether, so this issue (and the technical ones) just go away :)

**Comment by MODE83-PROJET-DEV** [comment]
@ditman 
Hi, I completely agree with you, but sorry I don't know how to transfer to flutter/flutter.

For the moment this works well in prod :
`
flutter build web --release --pwa-strategy none
`
If you have any other information, please do not hesitate.
kind regards

**Comment by johnpryan** [comment]
Thanks everyone for the discussion, I'm going to move this issue to flutter/flutter so we can track this feedback along with other service worker issues. The best option might be to remove the service worker from the framework so that developers can work with the service worker API as they see fit.
